src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -4928,6 +4925,53 @@ protected boolean isSingleByteOptimizable(AbstractTruffleString string, RubyEnco
         }
     }
 
+    @ImportStatic(TStringGuards.class)
+    public abstract static class GetCodePointNode extends RubyBaseNode {
+        @Child protected TruffleString.GetByteCodeRangeNode codeRangeNode = TruffleString.GetByteCodeRangeNode.create();
+
+        public static GetCodePointNode create() {
+            return StringNodesFactory.GetCodePointNodeGen.create();
+        }
+
+        public abstract int executeGetCodePoint(AbstractTruffleString string, RubyEncoding encoding, int byteIndex);
+
+        @Specialization(guards = "!isBrokenCodeRange(string, encoding, codeRangeNode)")
+        protected int getCodePointNonBroken(AbstractTruffleString string, RubyEncoding encoding, int byteIndex,
+                @Cached TruffleString.CodePointAtByteIndexNode getCodePointNode) {
+            return getCodePointNode.execute(string, byteIndex, encoding.tencoding);
+        }
+
+        @Specialization(guards = "isBrokenCodeRange(string, encoding, codeRangeNode)")
+        protected int getCodePointBroken(AbstractTruffleString string, RubyEncoding encoding, int byteIndex,
+                @Cached TruffleString.CodePointAtByteIndexNode getCodePointNode,
+                @Cached TruffleString.ByteLengthOfCodePointNode codePointLengthNode,
+                @Cached ConditionProfile potentiallyBadCodePointProfile,
+                @Cached ConditionProfile definitelyBadCodePointProfile) {
+            int codePoint = getCodePointNode.execute(string, byteIndex, encoding.tencoding);
+
+            // TruffleString will return the Unicode Replacement Character if a valid code point cannot be read.
+            // Of course, it will also return that code point if the string has that cade point at the specified byte index.
+            if (potentiallyBadCodePointProfile.profile(codePoint == 0xfffd)) {
+                int characterLength = codePointLengthNode.execute(string, byteIndex, encoding.tencoding);
+
+                // TruffleString will return a byte length of 1 for invalid code points. We can use that knowledge to
+                // distinguish whether a Unicode Replacement Character appeared in the string or was used as a default
+                // value for a broken byte range. If it appears in the string, it should have the Replacement Character's
+                // true byte length of 3.
```

Comment:
This is the case for UTF-8, but it's not always 1 e.g. for UTF-16 and others I believe:
```java
        @Specialization(guards = {"isUTF16(a)", "isBrokenMultiByte(codeRangeA)"})
        int utf16Broken(AbstractTruffleString a, Object arrayA, @SuppressWarnings("unused") int codeRangeA, int index) {
            assert isStride1(a);
            return Encodings.isUTF16HighSurrogate(TStringOps.readS1(a, arrayA, index)) && index + 1 < a.length() && Encodings.isUTF16LowSurrogate(TStringOps.readS1(a, arrayA, index + 1)) ? 2 : 1;
        }

        @TruffleBoundary
        @Specialization(guards = "isUnsupportedEncoding(a)")
        int unsupported(AbstractTruffleString a, Object arrayA, @SuppressWarnings("unused") int codeRangeA, int index) {
            JCodings.Encoding jCoding = JCodings.getInstance().get(a.encoding());
            int len = JCodings.getInstance().getCodePointLength(jCoding, JCodings.asByteArray(arrayA), a.byteArrayOffset() + index, a.byteArrayOffset() + a.length());
            return len < 0 ? JCodings.getInstance().minLength(jCoding) : len;
        }
```
I think the check is fine (because no encoding has minLength == 3), but we need to update the comment to reflect some of that.

Reply:
Of course, a variant of CodePointAtByteIndexNode (e.g., with an extra argument) which throws an exception (without stacktrace, probably a checked exception) instead of returning `0xfffd` would be very helpful, then we could just use a BranchProfile and avoid those extra checks. cc @djoooooe

Reply:
Fixed now with the new `ErrorHandling.RETURN_NEGATIVE`